core/src/test/java/com/linecorp/armeria/server/ServerTest.java

```diff
@@ -51,85 +52,89 @@
-                    @Override
-                    public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-                        ctx.setRequestTimeoutMillis(
-                                "/timeout-not".equals(ctx.path()) ? 0 : requestTimeoutMillis);
-                        return delegate().serve(ctx, req);
-                    }
-                };
+            final Service<HttpRequest, HttpResponse> lazyResponseNotOnIoThread = new EchoService() {
+                @Override
+                protected void echo(AggregatedHttpMessage aReq, HttpResponseWriter res) {
+                    asyncExecutorGroup.schedule(
+                            () -> super.echo(aReq, res), processDelayMillis, TimeUnit.MILLISECONDS);
+                }
+            }.decorate(LoggingService::new);
 
-        sb.decorator(decorator);
+            final Service<HttpRequest, HttpResponse> buggy = new AbstractHttpService() {
+                @Override
+                protected void doPost(ServiceRequestContext ctx,
+                                      HttpRequest req, HttpResponseWriter res) throws Exception {
 
-        sb.idleTimeoutMillis(idleTimeoutMillis);
-    }
+                    throw Exceptions.clearTrace(new Exception("bug!"));
+                }
+            }.decorate(LoggingService::new);
+
+            sb.serviceAt("/", immediateResponseOnIoThread)
+              .serviceAt("/delayed", delayedResponseOnIoThread)
+              .serviceAt("/timeout", lazyResponseNotOnIoThread)
+              .serviceAt("/timeout-not", lazyResponseNotOnIoThread)
+              .serviceAt("/buggy", buggy);
+
+            // Disable request timeout for '/timeout-not' only.
+            final Function<Service<HttpRequest, HttpResponse>, Service<HttpRequest, HttpResponse>> decorator =
+                    s -> new SimpleDecoratingService<HttpRequest, HttpResponse>(s) {
+                        @Override
+                        public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+                            ctx.setRequestTimeoutMillis(
+                                    "/timeout-not".equals(ctx.path()) ? 0 : requestTimeoutMillis);
+                            return delegate().serve(ctx, req);
+                        }
+                    };
+
+            sb.decorator(decorator);
+
+            sb.idleTimeoutMillis(idleTimeoutMillis);
+        }
+    };
 
     @Test
     public void testStartStop() throws Exception {
+        final Server server = ServerTest.server.server();
         try {
-            assertThat(server().activePorts().size(), is(1));
-            server().stop().get();
-            assertThat(server().activePorts().size(), is(0));
+            assertThat(server.activePorts().size(), is(1));
+            server.stop().get();
+            assertThat(server.activePorts().size(), is(0));
         } finally {
-            stopServer();
+            server.start();
```

Comment:
It actually attempts to start the server back because other test methods assumes the server is running. I agree this is confusing and potentially inefficient, so I just added a `@Before` method that ensures the server is started.